### PR TITLE
Add `tgs` speed metrics

### PR DESCRIPTION
**Pull request by CokeDong**
add tgs metrics for `trainer`. the motivation is that: current `speed_metrics` only consider `train_samples_per_second`. but the length of each example is  not the same(especailly `cutting_off ` increase). this pr introduce `tgs` metrics, which take `tokens` into considerations.

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @muellerzr @pacman100

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1154,6 +1154,22 @@ def num_examples(self, dataloader: DataLoader) -> int:
...
+        Helper to get number of tokens in a [`~torch.utils.data.DataLoader`] by enumerating dataloader.
+        """
+        try:
+            train_tokes = 0
+            for step, batch in enumerate(train_dl):
+                tks = batch["input_ids"].size(0) * batch["input_ids"].size(1)
+                if max_steps:
+                    return tks
+                train_tokes += tks
+            return train_tokes
```
Comment (muellerzr) [inline src/transformers/trainer.py:1168]: Is there a better way we can perhaps do this by checking if `max_steps` is none and its streaming dataloader for instance? I worry about speed slowdowns if users for instance have a large dataset across the dataloaders. 

I'd prefer this as perhaps an opt-in via `TrainingArguments`, noting how it will iterate over the dataloader to get these.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1154,6 +1154,22 @@ def num_examples(self, dataloader: DataLoader) -> int:
...
             return len(dataloader) * self.args.per_device_train_batch_size
 
+    def num_tokens(self, train_dl: DataLoader, max_steps: bool = False) -> int:
+        """
+        Helper to get number of tokens in a [`~torch.utils.data.DataLoader`] by enumerating dataloader.
+        """
+        try:
+            train_tokes = 0
+            for step, batch in enumerate(train_dl):
+                tks = batch["input_ids"].size(0) * batch["input_ids"].size(1)
```
Comment (muellerzr) [inline src/transformers/trainer.py:1164]:
```suggestion
                tokens = batch["input_ids"].size(0) * batch["input_ids"].size(1)
```

Full words please :)

**Review by muellerzr — COMMENTED** [review]
Thanks! having tokens per second would be quite handy, though I did make a nit on trying to save on time/streaming datasets. We may want this as an opt-in instead.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1588,17 +1604,30 @@ def _inner_training_loop(
...
+                    args.max_steps
+                    * self.num_tokens(train_dataloader, True)
+                    * args.gradient_accumulation_steps
+                    * args.world_size
+                )
             else:
                 max_steps = math.ceil(args.num_train_epochs * num_update_steps_per_epoch)
                 num_train_epochs = math.ceil(args.num_train_epochs)
                 num_train_samples = self.num_examples(train_dataloader) * args.num_train_epochs
+                num_train_tokens = self.num_tokens(train_dataloader) * args.world_size * args.num_train_epochs
```
Comment (pacman100) [inline src/transformers/trainer.py:1617]: We are calculating `max_steps` on line 1614. Why can't we use that to avoid iterating through the entire dataloader?

**Review by pacman100 — COMMENTED** [review]
Hello @CokeDong, Thank for you adding `tokens_per_second` log metric, helpful. Left a comment. I agree with Zach that iterating over the dataloader needs to be avoided as it will slow down the whole training.

File: src/transformers/trainer.py (hunk 2 above)
Comment (CokeDong) [reply src/transformers/trainer.py]: fixed.

File: src/transformers/trainer.py (hunk 3 above)
Comment (CokeDong) [reply src/transformers/trainer.py:1617]: `max_steps` doesnot contains the num of `tokens` infomation, so as call `num_tokens`.  

yes, the iteration will slow down the whole training ,especally big datasets. Currently i am trying moving the approximation of tokens per second into each training step so as to avoild duplicated dataloading process. maybe that will solve the common concerns.

Comment (muellerzr) [reply src/transformers/trainer.py:1617]: If that's not possible (which I don't think it is without some effort) I'd rather see this as an opt-in. Because the other issue with this is users can use the `Trainer` to train any dataset, not just text (though that's the most common). So again, I'd rather see this as an opt-in specifying more.

Comment (CokeDong) [reply src/transformers/trainer.py:1617]: yeah.i got u:-) an opt-in maybe the most direct way for user to choose open `tgs` or not.

**Comment by CokeDong** [comment]
@muellerzr @pacman100 PTAL, thx

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -631,6 +631,14 @@ class TrainingArguments:
...
 
             This flag is experimental and subject to change in future releases.
+        tgs_metrics (`bool`, *optional*, defaults to `False`):
+            Whether or not to compute the `tgs` for training speed metrics.
+
+            If set to `True`, the speed metrics will include `tgs`(tokens per second per device).
+
+            It will iterate over the dataloader ahead of training.
+
+            This will slow down the whole process, especally for large datasets.
```
Comment (muellerzr) [inline src/transformers/training_args.py:641]:
```suggestion
        include_tokens_per_second (`bool`, *optional*):
            Whether or not to compute the number of tokens per second per device for training speed metrics.
            This will iterate over the entire training dataloader once beforehand, and will slow down the entire process.
```

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -1230,6 +1238,11 @@ class TrainingArguments:
         },
     )
 
+    tgs_metrics: Optional[bool] = field(
```
Comment (muellerzr) [inline src/transformers/training_args.py:1241]:
```suggestion
    include_tokens_per_second: Optional[bool] = field(
```

**Review by muellerzr — COMMENTED** [review]
Thanks! We're getting much, much closer. I've added a naming nit, cc @amyeroberts if you have some ideas on other naming conventions, and how this looks to you :) Let's avoid acronyms/shorthand as much as possible.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1154,6 +1154,22 @@ def num_examples(self, dataloader: DataLoader) -> int:
...
+            train_tokens = 0
+            for step, batch in enumerate(train_dl):
+                tokens = batch["input_ids"].numel()
+                if max_steps:
+                    return tokens
+                train_tokens += tokens
+            return train_tokens
+        except KeyError:
+            logger.warning("Cannot get num_tokens from dataloader")
+            return 0
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1171]: I'd suggest changing the logic here slightly for three reasons
* `max_steps` in the rest of trainer is an int, rather than a bool
* What `max_steps` means isn't clearer (why this early return?)
* If `max_steps` is `True` then what's returned isn't the number of tokens (as indicated by the function name and docstring)

Instead, we can take `max_steps` as an int

```suggestion
    def num_tokens(self, train_dl: DataLoader, max_steps: Optional[int] = None) -> int:
        """
        Helper to get number of tokens in a [`~torch.utils.data.DataLoader`] by enumerating dataloader.
        """
        train_tokens = 0 
        try:
            for step, batch in enumerate(train_dl):
                tokens = batch["input_ids"].numel()
                if max_steps is not None:
                    return tokens * max_steps
                train_tokens += tokens
            return train_tokens
        except KeyError:
            logger.warning("Cannot get num_tokens from dataloader")
            return train_tokens
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for the work adding this @CokeDong! 

Overall code looks good - as @muellerzr suggests we should rename some of the variables, and slightly rework the logic to make the code clearer.

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -1575,6 +1591,7 @@ def _inner_training_loop(
         total_train_batch_size = self._train_batch_size * args.gradient_accumulation_steps * args.world_size
 
         len_dataloader = None
+        num_train_tokens = 0
```
[the remark this answers is not in the document]
Comment (CokeDong) [reply src/transformers/trainer.py:1594]: yep, refacted

**Comment by CokeDong** [comment]
https://app.circleci.com/jobs/github/huggingface/transformers/913233 `test_hub` failed， seems not related with current PR :(

**Comment by muellerzr** [comment]
@CokeDong please rebase with the main branch of transformers, this should ensure it passes :)

**Comment by muellerzr** [comment]
BTW @CokeDong, you don't have to do force-pushes if you're worried about the commit-bloat post-merge, in `transformers` we squash when merging.

**Comment by CokeDong** [comment]
> BTW @CokeDong, you don't have to do force-pushes if you're worried about the commit-bloat post-merge, in `transformers` we squash when merging. 

Got that

**Review by muellerzr — APPROVED** [review]
Thanks for iterating! LG2M

**Review by amyeroberts — APPROVED** [review]
Thanks for adding and iterating on this!

**Comment by geronimi73** [comment]
thanks for this great feature! 
quickly tried it with accelerate on 1-3 GPUs and the results confuse me 🤔 
t/s should be higher with 3 GPUs vs 1 GPU, right?

W B Chart 13_11_2023, 06_25_18
W B Chart 13_11_2023, 06_26_15

**Comment by CokeDong** [comment]
> thanks for this great feature! quickly tried it with accelerate on 1-3 GPUs and the results confuse me 🤔 t/s should be higher with 3 GPUs vs 1 GPU, right?
> 
> W B Chart 13_11_2023, 06_25_18 W B Chart 13_11_2023, 06_26_15

hi,  `tokens/sec/gpu`(tgs)  meatures tokens throughput capability per device.

**Comment by geronimi73** [comment]
> per device

now it makes sense, thank you!
